packages/loopring_v3/contracts/core/impl/libtransactions/WithdrawTransaction.sol

```diff
@@ -79,20 +85,29 @@ library WithdrawTransaction
         Withdrawal memory withdrawal = readWithdrawal(data, offset);
         WithdrawalAuxiliaryData memory auxData = abi.decode(auxiliaryData, (WithdrawalAuxiliaryData));
 
-        // Validate the auxixliary withdrawal data
-        if (withdrawal.dataHash == 0) {
-            require(auxData.auxiliaryData.length == 0, "AUXILIARY_DATA_NOT_ALLOWED");
-        } else {
-            // Hashes are stored using only 253 bits so the value fits inside a SNARK field element.
-            require(
-                uint(keccak256(auxData.auxiliaryData)) >> 3 == uint(withdrawal.dataHash),
-                "INVALID_WITHDRAWAL_AUX_DATA"
-            );
-        }
+        // Validate the withdrawal data not directly part of the DA
+        bytes32 dataHash = keccak256(
+            abi.encodePacked(
+                auxData.minGas,
+                auxData.to,
+                auxData.data
+            )
+        );
+        // Only the 20 MSB are used, which is still 80-bit of security, which is more
+        // than enough, especially when combined with validUntil.
+        require(withdrawal.dataHash == bytes20(dataHash), "INVALID_WITHDRAWAL_DATA");
+
+        // Fill in withdrawal data missing from DA
+        withdrawal.to = auxData.to;
+        withdrawal.minGas = auxData.minGas;
+        withdrawal.data = auxData.data;
+        withdrawal.validUntil = auxData.validUntil;
 
         if (withdrawal.withdrawalType == 0) {
             // Signature checked offchain, nothing to do
         } else if (withdrawal.withdrawalType == 1) {
+            // Check validUntil
+            require(block.timestamp < withdrawal.validUntil, "WITHDRAWAL_EXPIRED");
```

Comment:
I'm not sure if we should pass-in the block time as a parameter so for each block, we only read 'block.timestamp' once!

Reply:
Checked to make sure, this doesn't read from storage (which is the expensive one) and is practically free, cost is maybe 3-5 gas or something like that, hard to measure exactly because of that.